Fixes issue #619 imports-blacklist support regex

ydah opened the pull request:
Closes: #619 

This PR is "rule.imports-blacklist" supports regex and match any sub-packages.

## example

```
[rule.imports-blacklist]
arguments = ["github.com/full/match", "all/subpkg", "wc/**/between", "wc/backward/**"]
```

configuration would prevent following imports:

```go
import "github.com/full/match"
import "all/subpkg"
import "all/subpkg/pkg1"
import "all/subpkg/pkg1/pkg2"
import "wc/between"
import "wc/pkg1/between"
import "wc/pkg1/pkg2/between"
import "wc/backward"
import "wc/backward/pkg"
import "wc/backward/pkg/pkg1"
```

chavacava commented on RULES_DESCRIPTIONS.md:
````diff
@@ -432,7 +432,7 @@ Example:
 ```toml
 [imports-blacklist]
-  arguments =["crypto/md5", "crypto/sha1"]
+  arguments =["crypto/md5", "crypto/sha1", "crypto/**/pkix"]
````
You could also complete the documentation of the configuration by explaining that regexps are also accepted

chavacava reviewed (changes requested):
Thanks for the PR!
I've left some comments on the changes you propose.

chavacava commented:
Hi @ydah, I've tested the PR and I have a question.
If I want to warn when using the standard `"log"` package I assume the configuration should be
```toml
[rule.imports-blacklist]
    arguments=["log"]
```
But when I run `revive` on its code base I get:
```
revivelib/core.go:5:2: should not use the following blacklisted import: "log"
logging/logger.go:6:2: should not use the following blacklisted import: "log"
revivelib/core.go:11:2: should not use the following blacklisted import: "github.com/mgechev/revive/logging"
```
I was expecting only the first two warnings. The third one is pointing to a package I've didn't blacklist. 
What should be the configuration to only catch uses of `"log"`?

ydah commented:
@chavacava  Thank you so much for your review!
I updated this PR.

> I was expecting only the first two warnings. The third one is pointing to a package I've didn't blacklist.
What should be the configuration to only catch uses of "log"?

The third warning was an error in the code I added. Sorry.
I also thought that changing the behavior to include all submodules even when not explicitly indicated by a regular expression would be a destructive change in some cases in projects already using it, so I limited it to cases where the regular expression was explicitly specified.

```
[rule.imports-blacklist]
arguments = ["all/subpkg"]
```

configuration would prevent following imports:

```go
import "all/subpkg"
```

configuration wouldn't prevent following imports:

```go
import "all/subpkg/pkg1"
import "all/subpkg/pkg1/pkg2"
```

ydah commented on rule/imports-blacklist.go:
```diff
@@ -23,16 +25,22 @@ func (r *ImportsBlacklistRule) configure(arguments lint.Arguments) {
-				argStr = fmt.Sprintf(`%q`, argStr)
-			}
+			argStr = strings.Replace(argStr, "/**", `(\W|\w)*`, -1)
 			r.blacklist[argStr] = true
```
I've changed to a slice of compiled regular expressions.

ydah commented on rule/imports-blacklist.go:
```diff
@@ -23,16 +25,22 @@ func (r *ImportsBlacklistRule) configure(arguments lint.Arguments) {
-				argStr = fmt.Sprintf(`%q`, argStr)
-			}
+			argStr = strings.Replace(argStr, "/**", `(\W|\w)*`, -1)
 			r.blacklist[argStr] = true
```
I’ve modified it to compile regular expressions here.

chavacava commented:
@ydah thanks for the PR!

ydah commented:
@chavacava Thank you so much!
